pkg/storegateway/bucket.go

```diff
@@ -260,23 +270,38 @@ func NewBucketStore(
 	return s, nil
 }
 
+func (s *BucketStore) subservices() []services.Service {
+	return []services.Service{s.snapshotter}
+}
+
 func (s *BucketStore) start(context.Context) error {
 	return nil
 }
 
-func (s *BucketStore) stop(error) error {
-	return nil
+func (s *BucketStore) stop(err error) error {
+	subservices := s.subservices()
+
+	errs := multierror.New(err)
+	for _, svc := range subservices {
+		if err := services.StopAndAwaitTerminated(context.Background(), svc); err != nil {
+			errs.Add(fmt.Errorf("stop %T: %w", svc, err))
+		}
+	}
+	return errs.Err()
 }
 
 // RemoveBlocksAndClose remove all blocks from local disk and releases all resources associated with the BucketStore.
 func (s *BucketStore) RemoveBlocksAndClose() error {
-	removeBlocksErr := s.removeAllBlocks()
-
+	errs := multierror.New()
+	if err := services.StopAndAwaitTerminated(context.Background(), s); err != nil {
+		errs.Add(fmt.Errorf("stopping subservices: %w", err))
+	}
 	// Release other resources even if it failed to close some blocks.
-	s.snapshotter.Stop()
-	s.indexReaderPool.Close()
+	if err := s.removeAllBlocks(); err != nil {
+		errs.Add(fmt.Errorf("remove all blocks: %w", err))
+	}
```

Comment:
(nit) I think we can remove this comment. Now the wording feels confusing, because the order of actions is opposite to what it says.